Vekterli/safe bucket ownership handover time point

vekterli opened the pull request:
@hakonhall Please review. This adds a period of time after each distributor bucket ownership handover edge where mutations requiring timestamp assignments are bounced back to the client. This is in order to avoid the possibility of generating the same timestamp as a previous distributor whose bucket subset we now own. The time period is configurable in whole seconds based on expected worst-case cluster clock skew and the feature may be disabled entirely by setting the configured value to zero.

Read-only operations and mutating operations that don't require new timestamps (such as `RemoveLocation` are allowed through).

As discussed offline, this is a stop-gap solution until we have time to implement a solution which looks at the maximum timestamp across all buckets now owned by the distributor upon state change.

vekterli committed:
Expose whether a state change implies ownership transfer

vekterli committed:
Add computing, configuring and propagating of safe time point

vekterli committed:
Bounce mutations if safe time point not yet reached

The safe time point is when we assume that the distributor can no longer
generate time stamps that collide with those generated by distributors
that owned a subset of our buckets earlier.

vekterli committed:
Add end-to-end unit test for safe time rejection of feed

vekterli committed:
Clean up end-to-end unit test

vekterli committed:
Test that bounce metric not incremented when operation is accepted

vekterli committed:
Fix linebreak

vekterli committed:
Add copyright header

hakonhall reviewed (approved):
Just minor comments

vekterli committed:
Use STALE_TIMESTAMP instead of BUSY

Update DocumentProtocol error printing to include all known protocol
error codes.

vekterli committed:
Use explicit nullptr comparison
